tests/core/matrix.rs

```diff
@@ -1136,3 +1136,124 @@ fn omatrix_to_string() {
         (svec.to_string(), smatr.to_string())
     );
 }
+
+#[test]
+fn column_iteration() {
+    // dynamic matrix
+    let dmat = nalgebra::dmatrix![
+    13,14,15;
+    23,24,25;
+    33,34,35;
+    ];
+    let mut col_iter = dmat.column_iter();
+    assert_eq!(col_iter.next(), Some(dmat.column(0)));
+    assert_eq!(col_iter.next(), Some(dmat.column(1)));
+    assert_eq!(col_iter.next(), Some(dmat.column(2)));
+    assert_eq!(col_iter.next(), None);
+
+    // statically sized matrix
+    let smat: nalgebra::SMatrix<f64, 2, 2> = nalgebra::matrix![1.0, 2.0; 3.0, 4.0];
+    let mut col_iter = smat.column_iter();
+    assert_eq!(col_iter.next(), Some(smat.column(0)));
+    assert_eq!(col_iter.next(), Some(smat.column(1)));
+    assert_eq!(col_iter.next(), None);
+}
+
+#[test]
+fn column_iteration_mut() {
+    let mut dmat = nalgebra::dmatrix![
+    13,14,15;
+    23,24,25;
+    33,34,35;
+    ];
+    let mut cloned = dmat.clone();
+    let mut col_iter = dmat.column_iter_mut();
+    assert_eq!(col_iter.next(), Some(cloned.column_mut(0)));
+    assert_eq!(col_iter.next(), Some(cloned.column_mut(1)));
+    assert_eq!(col_iter.next(), Some(cloned.column_mut(2)));
+    assert_eq!(col_iter.next(), None);
+
+    // statically sized matrix
+    let mut smat: nalgebra::SMatrix<f64, 2, 2> = nalgebra::matrix![1.0, 2.0; 3.0, 4.0];
+    let mut cloned = smat.clone();
+    let mut col_iter = smat.column_iter_mut();
+    assert_eq!(col_iter.next(), Some(cloned.column_mut(0)));
+    assert_eq!(col_iter.next(), Some(cloned.column_mut(1)));
+    assert_eq!(col_iter.next(), None);
+}
+
+#[test]
+fn column_iteration_double_ended() {
+    let dmat = nalgebra::dmatrix![
+    13,14,15,16,17;
+    23,24,25,26,27;
+    33,34,35,36,37;
+    ];
+    let mut col_iter = dmat.column_iter();
+    assert_eq!(col_iter.next(), Some(dmat.column(0)));
+    assert_eq!(col_iter.next(), Some(dmat.column(1)));
+    assert_eq!(col_iter.next_back(), Some(dmat.column(4)));
+    assert_eq!(col_iter.next_back(), Some(dmat.column(3)));
+    assert_eq!(col_iter.next(), Some(dmat.column(2)));
+    assert_eq!(col_iter.next_back(), None);
+    assert_eq!(col_iter.next(), None);
+}
+
+#[test]
+#[cfg(feature = "rayon")]
+fn parallel_column_iteration() {
+    use nalgebra::dmatrix;
+    use rayon::prelude::*;
+    let dmat: DMatrix<f64> = dmatrix![
+    13.,14.;
+    23.,24.;
+    33.,34.;
+    ];
+    let cloned = dmat.clone();
+    // test that correct columns are iterated over
+    dmat.par_column_iter().enumerate().for_each(|(idx, col)| {
+        assert_eq!(col, cloned.column(idx));
+    });
+    // test that a more complex expression produces the same
+    // result as the serial equivalent
+    let par_result: f64 = dmat.par_column_iter().map(|col| col.norm()).sum();
+    let ser_result: f64 = dmat.column_iter().map(|col| col.norm()).sum();
+    assert_eq!(par_result, ser_result);
```

Comment:
Maybe repeat this test for the mutable case?

Reply:
I had to look it up myself, but I think I did that. The test is called `parallel_column_iteration_mut()` and is located around line 1250. Are you unhappy with the test for the mutable iteration?

Reply:
I had to look it up myself, but I think I did that. The test is called `parallel_column_iteration_mut()` and is located around line 1250. Are you unhappy with the test for the mutable iteration?

Reply:
That test appears to do something completely different, if I'm not mistaken?

Reply:
I mean, that test is also useful, but testing both behaviors seems like a good idea?

Reply:
ok sure, I'll add another test

Reply:
Hey thanks for the suggestion, I've appended the `parallel_column_iteration_mut` test to cover mutable iterators as well. That made me find a bug and I've added another test `column_iterator_double_ended_mut` that covers that bug.